paho/client.go

```diff
@@ -127,154 +138,167 @@ func NewClient(conf ClientConfig) *Client {
-		}
-		if cp.Properties.MaximumQOS != nil {
-			c.clientProps.MaximumQoS = *cp.Properties.MaximumQOS
+		keepalive := cp.KeepAlive
+		if keepalive == 0 {
+			keepalive = uint16(DefaultKeepAlive / time.Second)
 		}
-		if cp.Properties.ReceiveMaximum != nil {
-			c.clientProps.ReceiveMaximum = *cp.Properties.ReceiveMaximum
-		}
-		if cp.Properties.TopicAliasMaximum != nil {
-			c.clientProps.TopicAliasMaximum = *cp.Properties.TopicAliasMaximum
+		if cp.Properties != nil {
+			if cp.Properties.MaximumPacketSize != nil {
+				c.clientProps.MaximumPacketSize = *cp.Properties.MaximumPacketSize
+			}
+			if cp.Properties.MaximumQOS != nil {
+				c.clientProps.MaximumQoS = *cp.Properties.MaximumQOS
+			}
+			if cp.Properties.ReceiveMaximum != nil {
+				c.clientProps.ReceiveMaximum = *cp.Properties.ReceiveMaximum
+			}
+			if cp.Properties.TopicAliasMaximum != nil {
+				c.clientProps.TopicAliasMaximum = *cp.Properties.TopicAliasMaximum
+			}
 		}
-	}
 
-	debug.Println("starting writer")
-	c.workers.Add(1)
-	go func() {
-		defer c.workers.Done()
-		c.writer()
-	}()
+		go c.writer()
+		go c.reader()
 
-	debug.Println("starting reader")
-	c.workers.Add(1)
-	go func() {
-		defer c.workers.Done()
-		c.reader()
-	}()
+		connCtx, cf := context.WithTimeout(ctx, c.PacketTimeout)
+		defer cf()
+		c.caCtx = &caContext{connCtx, make(chan *packets.Connack, 1)}
+		defer func() {
+			c.caCtx = nil
+		}()
 
-	connCtx, cf := context.WithTimeout(ctx, c.PacketTimeout)
-	defer cf()
-	c.caCtx = &caContext{connCtx, make(chan *packets.Connack, 1)}
-	defer func() {
-		c.caCtx = nil
-	}()
-
-	ccp := cp.Packet()
+		ccp := cp.Packet()
+		ccp.ProtocolName = "MQTT"
+		ccp.ProtocolVersion = 5
 
-	ccp.ProtocolName = "MQTT"
-	ccp.ProtocolVersion = 5
+		debug.Println("sending CONNECT")
+		if c.cerr = c.write(ctx, ccp); c.cerr != nil {
+			return
+		}
 
-	debug.Println("sending CONNECT")
-	if err := c.write(ctx, ccp); err != nil {
-		cleanup()
-		return nil, err
-	}
+		debug.Println("waiting for CONNACK")
+		var cap *packets.Connack
+		select {
+		case <-connCtx.Done():
+			debug.Println("waiting for CONNACK canceled")
+			c.cerr = connCtx.Err()
+			return
 
-	debug.Println("waiting for CONNACK")
-	var cap *packets.Connack
-	select {
-	case <-connCtx.Done():
-		debug.Println("waiting for CONNACK canceled")
-		cleanup()
-		return nil, connCtx.Err()
-	case cap = <-c.caCtx.Return:
-	}
+		case cap = <-c.caCtx.Return:
+			debug.Println("received CONNACK")
+		}
 
-	ca := ConnackFromPacketConnack(cap)
+		ca := ConnackFromPacketConnack(cap)
 
-	if ca.ReasonCode >= 0x80 {
-		var reason string
-		debug.Println("received an error code in Connack:", ca.ReasonCode)
-		if ca.Properties != nil {
-			reason = ca.Properties.ReasonString
+		if ca.ReasonCode >= 0x80 {
+			var reason string
+			debug.Println("received an error code in Connack:", ca.ReasonCode)
+			if ca.Properties != nil {
+				reason = ca.Properties.ReasonString
+			}
+			c.cerr = fmt.Errorf("failed to connect to server: %s", reason)
```

Comment:
So the user will need to create a a new object before trying to connect again, as this one will have used up its `sync.Once`? What about reconnection in general?

Reply:
Yes. The code would much more complicated if we want to deal with cleaning up all started goroutines or closed channels.

Reply:
ok